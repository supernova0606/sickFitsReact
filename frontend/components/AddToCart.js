import { Component } from 'react';
import { graphql, compose } from 'react-apollo';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { removeFromCartEnhancer, userEnhancer, addToCartEnhancer, singleItemEnhancer } from '../enhancers/enhancers';

class AddToCart extends Component {
  static propTypes = {
    currentUser: PropTypes.object.isRequired,
  };

  componentDidMount() {
    this.props.currentUser.refetch();
  }

  handleAddToCart = async () => {
    console.log(`Gonna add this item to the cart! ${this.props.id}`);
    const res = await this.props.addToCart({
      variables: {
        id: this.props.id,
      },
    });
    // this.props.currentUser.refetch();
  };

  render() {
    return <button onClick={this.handleAddToCart}>ðŸ›’ Add To Cart</button>;
  }
}

export default compose(userEnhancer, addToCartEnhancer)(AddToCart);
