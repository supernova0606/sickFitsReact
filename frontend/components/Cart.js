import { Component } from 'react';
import { graphql, compose } from 'react-apollo';
import styled from 'styled-components';
import { CURRENT_USER_QUERY } from '../queries';
import RemoveFromCart from './RemoveFromCart';
import TakeMyMoney from './TakeMyMoney';
import { formatMoney } from '../lib/formatMoney';

const CartStyles = styled.div`
  background: white;
  border-radius: 10px;
  padding: 20px;
  overflow: hidden;
  display: flex;
  align-items: center;
`;

class Cart extends Component {
  componentDidMount() {
    console.log('refetching..');
    setTimeout(this.props.currentUser.refetch, 10);
  }
  render() {
    if (!this.props.currentUser.me) {
      return null;
    }
    // Check for loading state..
    // const { loading, error } = this.props.currentUserQuery;
    // const { user } = this.props.currentUserQuery;
    // if (loading || error || !user) return <p>Cart Loading...</p>;
    // const total = user.cart.reduce((a, b) => a + b.price, 0);

    const { me } = this.props.currentUser;
    console.log(me);
    return (
      <CartStyles className="cart">
        <p>ðŸ›’: {me.cart.length} Items</p>
        <ul>
          {me.cart.map(cartItem => (
            <li key={cartItem.id}>
              <strong>
                {cartItem.quantity} of {cartItem.item.title}
              </strong>
              <RemoveFromCart id={cartItem.id} />
            </li>
          ))}
        </ul>

        <TakeMyMoney>
          <hr />
          <button>Checkout!!!</button>
        </TakeMyMoney>
        {/* There
        {user.cart.length === 1 ? ' is ' : 'are '}
        <ChaChing amount={user.cart.length} />
        {user.cart.length === 1 ? ' item ' : ' items '}
        in your cart totaling */}
        {/* <ChaChing amount={formatMoney(total)} /> */}
      </CartStyles>
    );
  }
}

const userEnhancer = graphql(CURRENT_USER_QUERY, { name: 'currentUser' });
export default compose(userEnhancer)(Cart);
