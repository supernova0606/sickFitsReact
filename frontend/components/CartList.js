import { Component } from 'react';
import withData from '../lib/withData';
import Items from '../components/Items';
import Signup from '../components/Signup';
import LoginAuth0 from '../components/LoginAuth0';
import Page from '../components/Page';
import { USER_ORDERS_QUERY, GET_CART_STATE } from '../queries';
import { graphql, compose } from 'react-apollo';
import has from 'lodash.has';
import get from 'lodash.get';
import styled from 'styled-components';
import formatMoney from '../lib/formatMoney';
import makeImage from '../lib/image';
import TakeMyMoney from './TakeMyMoney';
import { removeFromCartEnhancer } from '../enhancers/enhancers';
import { CURRENT_USER_QUERY } from '../queries';
import PropTypes from 'prop-types';

const CartStyles = styled.div`
  padding: 20px;
  position: relative;
  background: white;
  position: fixed;
  height: 100%;
  top: 0;
  right: 0;
  bottom: 0;
  transform: translateX(100%);
  transition: all 0.3s;
  box-shadow: 0 0 10px 3px rgba(0, 0, 0, 0.2);
  z-index: 5;
  ${props => props.open && `transform: translateX(0);`};
  .toggleOpen {
    position: absolute;
    top: 0;
    left: 0;
    transform: translateX(-100%);
  }
`;

class CartList extends Component {
  static contextTypes = {
    client: PropTypes.object,
  };

  componentDidMount() {
    setTimeout(this.props.currentUserQuery.refetch, 1);
    // Set the cart state to be closed
    this.context.client.writeQuery({
      query: GET_CART_STATE,
      data: { ui: { isCartOpen: true, __typename: 'Network' } },
    });
  }

  toggleOpen = () => {
    this.props.uiQuery.updateQuery(() => ({ ui: { isCartOpen: false, __typename: 'Network' } }));
  };

  render() {
    if (this.props.loading) {
      return <p>Loading....</p>;
    }

    if (this.props.error) {
      return <p>Error....</p>;
    }

    // if (!has(this.props, 'currentUserQuery.user.cart')) {
    //   return <p>Don't have it yet!</p>;
    // }
    // OMG Clean this up
    let { user = {} } = this.props.currentUserQuery || {};
    if (!user) user = {};
    const cart = user.cart || [];
    const userId = user.id;
    // const userId = this.props.currentUserQuery.user.id;

    const total = cart.reduce((a, b) => a + b.price, 0);
    return (
      <CartStyles open>
        <button className="toggleOpen" onClick={this.toggleOpen}>
          Open Cart
        </button>
        <h1>{cart.length} Items</h1>

        <ul>
          {cart.map(item => (
            <li key={item.id}>
              {item.title}
              <button
                onClick={() =>
                  this.props.removeFromCart({
                    variables: {
                      userId,
                      itemId: item.id,
                    },
                  })}
              >
                &times; Delete
              </button>
            </li>
          ))}
        </ul>
        <TakeMyMoney amount={total} name="Testing 123" description="Test test 123">
          <button>Buy for {formatMoney(total)}</button>
        </TakeMyMoney>
      </CartStyles>
    );
  }
}

const userEnhancer = graphql(CURRENT_USER_QUERY, { name: 'currentUserQuery' });
const uiEnhancer = graphql(GET_CART_STATE, { name: 'uiQuery' });
export default compose(userEnhancer, removeFromCartEnhancer, uiEnhancer)(CartList);
