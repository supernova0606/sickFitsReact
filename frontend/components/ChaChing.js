import TransitionGroup from 'react-transition-group/TransitionGroup';
import Transition from 'react-transition-group/Transition';
import styled from 'styled-components';

const CartCount = styled.div`
  padding: 10px;
  transition: all 0.5s;
  display: inline-block;
  background: white;
  background: linear-gradient(to bottom, #eeeeee 0%, #eeeeee 100%);
  transform: translateY(100%);
  z-index: 1;
  position: relative;
  &.cart-entering {
    background: green;
    transform: translateY(100%);
  }
  &.cart-entered {
    transform: translateY(0);
  }
  &.cart-exiting {
    position: absolute;
    z-index: 0;
    left: 0;
    transform: translateY(-100%);
  }
`;

const ChaChingStyles = styled.div`
  position: relative;
  display: inline-block;
  overflow: hidden;
`;

const ChaChing = ({ amount }) => (
  <ChaChingStyles>
    <TransitionGroup>
      <Transition
        in
        timeout={{
          enter: 0,
          exit: 500,
        }}
        key={amount}
      >
        {status => (
          <CartCount className={`cart-${status}`} key={`count-${amount}`}>
            {amount}
          </CartCount>
        )}
      </Transition>
    </TransitionGroup>
  </ChaChingStyles>
);

export default ChaChing;
