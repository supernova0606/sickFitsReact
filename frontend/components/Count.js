// TODO THis is not needed
import React, { Component } from 'react'
import { graphql, gql } from 'react-apollo'

import { ALL_ITEMS_QUERY } from '../queries';

class Count extends Component {

  render() {

    // 1
    if (this.props.allLinksQuery && this.props.allLinksQuery.loading) {
      return <div>Loading</div>
    }

    // 2
    if (this.props.allLinksQuery && this.props.allLinksQuery.error) {
      return <div>Error</div>
    }

    // 3
    const itemsToRender = this.props.allLinksQuery.allItems

    return (
      <div>
        <h1>There are {itemsToRender.length} items for sale</h1>
      </div>
    )
  }

}

// 1

// We export the graphQL HOC - this will fetch the data and inject it into the Count compeont via props

export default graphql(ALL_ITEMS_QUERY, { name: 'allLinksQuery' })(Count)
