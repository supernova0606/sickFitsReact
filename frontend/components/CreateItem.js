import React, { Component } from 'react';
import { graphql, gql } from 'react-apollo';
import { CREATE_ITEM_MUTATION } from '../queries';
import ErrorMessage from './ErrorMessage';
import Form from './styles/Form';
import PropTypes from 'prop-types';

class CreateItem extends Component {
  static propTypes = {
    createItemMutation: PropTypes.func.isRequired,
  };

  state = {
    item: {
      title: '',
      description: '',
      image: '',
      largeImage: '',
      price: 0,
    },
    loading: false,
    error: {
      message: null,
    },
  };

  handleChange = e => {
    const { name, value, type } = e.target;
    const updatedItem = {
      ...this.state.item,
      [name]: type === 'number' ? parseFloat(value) : value,
    };
    this.setState({ item: updatedItem });
  };

  uploadFile = async e => {
    this.setState({ loading: true });
    const files = e.currentTarget.files;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'sickfits');

    // use the file endpoint
    const res = await fetch('https://api.cloudinary.com/v1_1/wesbos/image/upload', {
      method: 'POST',
      body: data,
    });
    const file = await res.json();
    this.setState({ image: file.secure_url, largeImage: file.eager[0].secure_url, loading: false });
  };

  createItem = async e => {
    e.preventDefault();
    // TODO: handle any errors
    // turn loading on
    this.setState({ loading: true });
    try {
      const res = await this.props.createItemMutation({
        // pass in those variables from state
        variables: {
          ...this.state.item,
        },
      });
    } catch (error) {
      this.setState({ error });
    }
    this.setState({ loading: false });
  };

  render() {
    return (
      <div>
        <Form onSubmit={this.createItem}>
          {this.state.loading ? 'LOADING...' : null}
          <ErrorMessage error={this.state.error} onButtonClick={() => this.setState({ error: {} })} />
          <p>
            Image
            <input onChange={this.uploadFile} type="file" accept=".png, .jpg, .jpeg" />
            {this.state.image ? <img src={this.state.image} width="100" alt={this.state.title} /> : null}
          </p>
          <p>
            Title
            <input
              value={this.state.title}
              onChange={this.handleChange}
              type="text"
              name="title"
              id="title"
              placeholder="Title"
            />
          </p>
          <label>
            Price
            <input
              type="number"
              id="price"
              name="price"
              min="0"
              value={this.state.price}
              onChange={this.handleChange}
            />
          </label>
          <textarea
            id="description"
            name="description"
            value={this.state.description}
            onChange={this.handleChange}
            placeholder="The desc for this item"
          />
          <button disabled={this.state.loading} type="submit">
            Submit
          </button>
        </Form>
      </div>
    );
  }
}

export default graphql(CREATE_ITEM_MUTATION, {
  name: 'createItemMutation',
  options: {
    refetchQueries: ['AllItemsQuery'],
  },
})(CreateItem);

export { CreateItem };
