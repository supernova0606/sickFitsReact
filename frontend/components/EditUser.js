import React from 'react';
import Form from './styles/Form';
import { userEnhancer, updateUserEnhancer } from '../enhancers/enhancers';
import { compose } from 'react-apollo';

class EditUser extends React.Component {
  state = {
    changes: {},
  };
  handleChange = e => {
    const { name, value } = e.target;
    const { me } = this.props.currentUser;
    const changes = {
      ...this.state.changes,
      [name]: value,
    };
    if (value === me[e.target.name]) {
      delete changes[name];
    }
    this.setState({ changes });
  };

  handleSubmit = async e => {
    e.preventDefault();
    const user = await this.props.updateUser({
      variables: this.state.changes,
    });
    this.props.currentUser.refetch();
    this.setState({ changes: {} });
  };

  render() {
    const { me } = this.props.currentUser;
    if (!me) return <p>You must be logged in</p>;
    return (
      <Form onSubmit={this.handleSubmit}>
        <label htmlFor="name">
          Name:
          <input type="text" name="name" defaultValue={me.name} onChange={this.handleChange} />
        </label>
        <button type="submit">Update</button>
        <strong>me:</strong>
        <pre>{JSON.stringify(me.name)}</pre>
        <strong>Change:</strong>
        <pre>{JSON.stringify(this.state.changes)}</pre>
      </Form>
    );
  }
}

export default compose(userEnhancer, updateUserEnhancer)(EditUser);
