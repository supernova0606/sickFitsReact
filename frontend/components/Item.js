import React from 'react';
import Title from './styles/Title';
import styled from 'styled-components';
import slugify from 'slugify';
import { compose } from 'react-apollo';
// import { Link } from '../routes';
import Link from 'next/link';
import AddToCart from './AddToCart';
import formatMoney from '../lib/formatMoney';
import { removeItemMutation } from '../enhancers/enhancers';

const Item = styled.div`
  background: white;
  border: 1px solid ${props => props.theme.offWhite};
  box-shadow: 0 12px 24px 0 rgba(0, 0, 0, 0.09);
  position: relative;
  display: grid;
  align-content: start;
  /* grid-template-rows: repeat(auto-fit, minmax(100px,)); */
  grid-auto-rows: fit-content;
  img {
    width: 100%;
  }
  p {
    font-size: 14px;
    line-height: 2;
    font-weight: 600;
    padding: 0 3rem;
    font-size: 1.5rem;
  }
  .buttonList {
    display: grid;
    border-top: 1px solid ${props => props.theme.lightgrey};
    grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));
    grid-gap: 1px;
    background: ${props => props.theme.lightgrey};
    align-self: end;
    & > * {
      background: white;
      border: 0;
      text-align: center;
      font-size: 1rem;
      padding: 1rem;
    }
  }
`;

const PriceTag = styled.span`
  background: ${props => props.theme.red};
  transform: rotate(2deg);
  color: white;
  font-weight: 600;
  padding: 5px;
  font-size: 3rem;
  display: inline-block;
  position: absolute;
  top: -3px;
  right: -3px;
`;

class ItemComponent extends React.Component {
  removeItem = () => {
    if (confirm('Are you sure you want to delete this item?')) {
      this.props.removeItem({ variables: { id: this.props.item.id } });
    }
  };
  render() {
    const item = this.props.item;
    return (
      <Item key={item.id}>
        {item.image ? <img src={item.image} alt={item.title} /> : null}
        <Title>
          <Link
            href={{
              pathname: '/item',
              query: { id: item.id },
            }}
          >
            <a>{item.title}</a>
          </Link>
        </Title>

        <PriceTag>{formatMoney(item.price)}</PriceTag>

        <p>{item.description}</p>

        <div className="buttonList">
          <Link
            href={{
              pathname: '/admin/update',
              query: { id: item.id },
            }}
          >
            <a>Edit ✏️</a>
          </Link>
          <AddToCart id={item.id} />
          <button onClick={this.removeItem}>&times; Delete item</button>
        </div>
      </Item>
    );
  }
}

export default compose(removeItemMutation)(ItemComponent);
