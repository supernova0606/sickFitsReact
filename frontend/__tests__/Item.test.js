/* eslint-env jest */

import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { ItemComponent } from '../components/Item';

const fakeItem = {
  id: 'ABC123',
  title: 'A Cool Item',
  price: 5000,
  description: 'This item is really cool!',
  image: 'dog.jpg',
  largeImage: 'largedog.jpg',
};

describe('<Item/>', () => {
  it('Renders an item', () => {
    const removeItem = jest.fn();
    const wrapper = shallow(<ItemComponent item={fakeItem} removeItem={removeItem} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
  it('handles button clicks', async () => {
    const removeItem = jest.fn();
    global.confirm = jest.fn().mockReturnValue(true);
    const wrapper = shallow(<ItemComponent item={fakeItem} removeItem={removeItem} />);
    wrapper.find('button').simulate('click');
    expect(removeItem).toHaveBeenCalled();
    expect(global.confirm).toHaveBeenCalled();
  });
});
