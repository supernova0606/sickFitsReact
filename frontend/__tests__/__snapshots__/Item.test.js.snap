// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item/> Renders an item 1`] = `
<styled.div
  key="ABC123"
>
  <img
    alt="A Cool Item"
    src="dog.jpg"
  />
  <styled.h3>
    <Link
      href={
        Object {
          "pathname": "/item",
          "query": Object {
            "id": "ABC123",
          },
        }
      }
    >
      <a>
        A Cool Item
      </a>
    </Link>
  </styled.h3>
  <styled.span>
    $50
  </styled.span>
  <p>
    This item is really cool!
  </p>
  <div
    className="buttonList"
  >
    <Link
      href={
        Object {
          "pathname": "/admin/update",
          "query": Object {
            "id": "ABC123",
          },
        }
      }
    >
      <a>
        Edit ✏️
      </a>
    </Link>
    <Apollo(Apollo(AddToCart))
      id="ABC123"
    />
    <button
      onClick={[Function]}
    >
      × Delete item
    </button>
  </div>
</styled.div>
`;
